File: web/src/app/position.ts
import { Engine } from "./engine";
import { Indicator } from "./indicator";

export interface Position {
  id: number;
  market: string;
  strategy_id: number;
  status: string;
  side: string;
  target_amount: number;
  target_cost: number;
  entry_cost: number;
  entry_amount: number;
  entry_time: Date;
  entry_price: number;
  exit_price: number;
  exit_time: Date | null;
  exit_amount: number;
  exit_cost: number;
  fee: number;
  pnl: number;
  roi: number;
  stopped: boolean;

  last_high?: number;
  last_low?: number;

  balance?: number;
  max_equity?: number;
  min_equity?: number;
  drawdown?: number;
}

export function getLongPositions(indicators: Indicator, stopEngine: Engine | null): Position[] {
  let init_balance = 100;
  let balance = init_balance;
  let fee = 0.001;

  let positions = [];

  let curr = false;
  let pos = {
    id: 0,
    market: "",
    strategy_id: 0,
    status: "closed",
    side: "long",
    target_amount: 0,
    target_cost: 0,
    entry_cost: 0,
    entry_amount: 0,
    entry_time: new Date(0),
    entry_price: 0,
    exit_price: 0,
    exit_time: new Date(0),
    exit_amount: 0,
    exit_cost: 0,
    fee: 0,
    pnl: 0,
    roi: 0,
    stopped: false,

    last_high: 0,
    last_low: 0,

    balance: init_balance,
    max_equity: 0,
    min_equity: 0,
    drawdown: 0,
  };

  let maxDrawdown = 0;

  for (let i = 0; i < indicators.time.length; i++) {
    let stopPrice = checkStop(pos, stopEngine, indicators.high[i], indicators.low[i]);
    let stopped = stopPrice > 0;

    let cooledDown = checkStopCooldown(pos, stopEngine, indicators.time[i]);

    if (indicators.buys[i] > 0 && !cooledDown) {
      if (!curr) {
        curr = true;

        maxDrawdown = 0;

        pos.entry_time = new Date(indicators.time[i]);
        pos.entry_price = indicators.close[i];

        pos.entry_amount = balance / pos.entry_price;
        pos.entry_cost = pos.entry_amount * pos.entry_price;
        pos.exit_amount = pos.entry_amount

        pos.max_equity = pos.entry_cost;
        pos.min_equity = pos.entry_cost;

        pos.last_high = pos.entry_price;
        pos.last_low = pos.entry_price;

      }
    } else if (indicators.sells[i] > 0 || stopped) {
      if (curr) {
        curr = false;

        pos.exit_price = indicators.close[i];
        if (stopped) {
          pos.exit_price = stopPrice;
        }

        pos.exit_time = new Date(indicators.time[i]);
        pos.exit_cost = pos.exit_amount * pos.exit_price;

        pos.fee = pos.entry_cost * fee + pos.exit_cost * fee;

        pos.pnl = pos.exit_cost - pos.entry_cost - pos.fee;
        pos.roi = (pos.pnl / pos.entry_cost) * 100;

        balance = balance + pos.pnl;
        pos.balance = balance;

        pos.drawdown = maxDrawdown;

        pos.stopped = stopped;

        positions.push({ ...pos });
      }
    }

    if (pos.entry_amount > 0) {
      let currMaxEquity = indicators.high[i] * pos.entry_amount;
      let currMinEquity = indicators.low[i] * pos.entry_amount;

      let currDrawdown = (currMinEquity - pos.max_equity) / pos.max_equity * 100;
      if (currDrawdown < maxDrawdown) {
        maxDrawdown = currDrawdown;
      }

      if (currMaxEquity > pos.max_equity) {
        pos.max_equity = currMaxEquity;
        pos.min_equity = currMaxEquity;
      }

      if (currMinEquity < pos.min_equity) {
        pos.min_equity = currMinEquity;
      }

    }
  }

  return positions;
}

export function getShortPositions(indicators: Indicator, stopEngine: Engine | null): Position[] {
  let init_balance = 100;
  let balance = init_balance;
  let fee = 0.001;

  let positions = [];

  let curr = false;
  let pos = {
    id: 0,
    market: "",
    strategy_id: 0,
    status: "closed",
    side: "short",
    target_amount: 0,
    target_cost: 0,
    entry_cost: 0,
    entry_amount: 0,
    entry_time: new Date(0),
    entry_price: 0,
    exit_price: 0,
    exit_time: new Date(0),
    exit_amount: 0,
    exit_cost: 0,
    fee: 0,
    pnl: 0,
    roi: 0,
    stopped: false,

    last_high: 0,
    last_low: 0,

    balance: init_balance,
    max_equity: 0,
    min_equity: 0,
    drawdown: 0,
  };

  let maxDrawdown = 0;

  for (let i = 0; i < indicators.time.length; i++) {
    let stopPrice = checkStop(pos, stopEngine, indicators.high[i], indicators.low[i]);
    let stopped = stopPrice > 0;

    let cooledDown = checkStopCooldown(pos, stopEngine, indicators.time[i]);

    if (indicators.sells[i] > 0 && !cooledDown) {
      if (!curr) {
        curr = true;

        maxDrawdown = 0;

        pos.entry_time = new Date(indicators.time[i]);
        pos.entry_price = indicators.close[i];

        pos.entry_amount = balance;
        pos.entry_cost = pos.entry_amount * pos.entry_price;
        pos.exit_cost = pos.entry_cost

        pos.max_equity = pos.entry_cost;
        pos.min_equity = pos.entry_cost;

        pos.last_high = pos.entry_price;
        pos.last_low = pos.entry_price;

      }
    } else if (indicators.buys[i] > 0 || stopped) {
      if (curr) {
        curr = false;

        pos.exit_price = indicators.close[i];
        if (stopped) {
          pos.exit_price = stopPrice;
        }

        pos.exit_time = new Date(indicators.time[i]);
        pos.exit_amount = pos.exit_cost / pos.exit_price;

        pos.fee = pos.entry_amount * fee + pos.exit_amount * fee;

        pos.pnl = pos.exit_amount - pos.entry_amount - pos.fee;
        pos.roi = (pos.pnl / pos.entry_amount) * 100;

        balance = balance + pos.pnl;
        pos.balance = balance;

        pos.drawdown = maxDrawdown;

        pos.stopped = stopped;

        positions.push({ ...pos });
      }
    }

    if (pos.entry_amount > 0) {
      let currMaxEquity = indicators.high[i] * pos.entry_amount;
      let currMinEquity = indicators.low[i] * pos.entry_amount;

      let currDrawdown = (currMinEquity - pos.max_equity) / pos.max_equity * 100;
      if (currDrawdown < maxDrawdown) {
        maxDrawdown = currDrawdown;
      }

      if (currMaxEquity > pos.max_equity) {
        pos.max_equity = currMaxEquity;
        pos.min_equity = currMaxEquity;
      }

      if (currMinEquity < pos.min_equity) {
        pos.min_equity = currMinEquity;
      }

    }
  }

  return positions;
}

function checkStop(pos: Position, engine: Engine | null, high: number, low: number): number {
  if (pos.entry_price == 0) {
    return 0;
  }

  if (engine === null) {
    return 0;
  }

  if (engine.stop_gain <= 0 && engine.stop_loss <= 0) {
    return 0;
  }

  if (high > pos.last_high!) {
    pos.last_high = high;
  }

  if (low < pos.last_low!) {
    pos.last_low = low;
  }

  if (engine.stop_gain > 0) {
    let currGain = 0;

    if (engine.trailing_gain) {

      if (low > pos.entry_price) {
        currGain = ((pos.last_high! - low) / low * 100) * -1;
        if (currGain > engine.stop_gain) {
          return pos.last_high! * (1 - engine.stop_gain / 100);
        }
      }

    } else {
      currGain = (high - pos.entry_price) / pos.entry_price * 100;
      if (currGain > engine.stop_gain) {
        return pos.entry_price * (1 + engine.stop_gain / 100);
      }
    }

  }

  if (engine.stop_loss > 0) {
    let currLoss = 0;

    if (engine.trailing_loss) {

      if (high < pos.entry_price) {
        currLoss = (pos.last_low! - high) / high * 100;
        if (currLoss > engine.stop_loss) {
          return pos.last_low! * (1 + engine.stop_loss / 100);
        }
      }

    } else {
      currLoss = ((low - pos.entry_price) / pos.entry_price * 100) * -1;
      if (currLoss > engine.stop_loss) {
        return pos.entry_price * (1 - engine.stop_loss / 100);
      }
    }

  }

  return 0;
}

function checkStopCooldown(pos: Position, engine: Engine | null, time: Date): boolean {
  let cooldown_in_min = engine?.stop_cooldown || 0;

  if (cooldown_in_min <= 0 || !pos.stopped) {
    return false;
  }

  let now = new Date(time);
  let last = pos.exit_time!;

  let diff_in_min = (now.getTime() - last.getTime()) / 1000 / 60;

  return diff_in_min < cooldown_in_min;
}

File: src/sliver/api/resources/position.py
from decimal import Decimal as D

import pandas
from flask_jwt_extended import get_jwt_identity, jwt_required
from flask_restful import Resource, fields, marshal_with

from sliver.api.exceptions import PositionDoesNotExist, StrategyDoesNotExist
from sliver.position import Position as PositionModel
from sliver.price import Price
from sliver.strategy import BaseStrategy as Strategy
from sliver.user import User
from sliver.utils import quantize

pos_fields = {
    "id": fields.Integer,
    "market": fields.String,
    "strategy_id": fields.Integer,
    "status": fields.String,
    "side": fields.String,
    "target_amount": fields.Float,
    "target_cost": fields.Float,
    "entry_cost": fields.Float,
    "entry_amount": fields.Float,
    "entry_price": fields.Float,
    "exit_price": fields.Float,
    "exit_amount": fields.Float,
    "exit_cost": fields.Float,
    "fee": fields.Float,
    "pnl": fields.Float,
    "roi": fields.Float,
    "stopped": fields.Boolean,
    "entry_time": fields.DateTime(dt_format="iso8601"),
    "exit_time": fields.DateTime(dt_format="iso8601"),
}

by_strategy_fields = {
    **pos_fields,
    "max_equity": fields.Float,
    "min_equity": fields.Float,
    "drawdown": fields.Float,
}


def get_positions_df(query):
    positions = pandas.DataFrame(query.dicts())

    if positions.empty:
        return positions

    positions["market"] = positions.base_ticker + "/" + positions.quote_ticker

    base_precision = D("10") ** (D("-1") * positions.base_precision)
    quote_precision = D("10") ** (D("-1") * positions.quote_precision)
    pnl_precision = D("10") ** (D("-1") * positions.pnl_precision)

    positions.target_amount = positions.target_amount * quote_precision
    positions.target_amount = positions.apply(
        lambda x: quantize(x, "target_amount", "amount_precision"), axis=1
    )

    positions.target_cost = positions.target_cost * quote_precision
    positions.target_cost = positions.apply(
        lambda x: quantize(x, "target_cost", "price_precision"), axis=1
    )

    positions.entry_cost = positions.entry_cost * quote_precision
    positions.entry_cost = positions.apply(
        lambda x: quantize(x, "entry_cost", "price_precision"), axis=1
    )

    positions.entry_amount = positions.entry_amount * base_precision
    positions.entry_amount = positions.apply(
        lambda x: quantize(x, "entry_amount", "amount_precision"), axis=1
    )

    positions.entry_price = positions.entry_price * quote_precision
    positions.entry_price = positions.apply(
        lambda x: quantize(x, "entry_price", "price_precision"), axis=1
    )

    positions.exit_price = positions.exit_price * quote_precision
    positions.exit_price = positions.apply(
        lambda x: quantize(x, "exit_price", "price_precision"), axis=1
    )

    positions.exit_amount = positions.exit_amount * base_precision
    positions.exit_amount = positions.apply(
        lambda x: quantize(x, "exit_amount", "amount_precision"), axis=1
    )

    positions.exit_cost = positions.exit_cost * quote_precision
    positions.exit_cost = positions.apply(
        lambda x: quantize(x, "exit_cost", "price_precision"), axis=1
    )

    positions.fee = positions.fee * quote_precision
    positions.fee = positions.apply(
        lambda x: quantize(x, "fee", "price_precision"), axis=1
    )

    positions.pnl = positions.pnl * pnl_precision
    positions.pnl = positions.apply(
        lambda x: quantize(
            x, "pnl", "price_precision" if x.side == "long" else "amount_precision"
        ),
        axis=1,
    )

    positions.roi = positions.roi.apply(lambda x: f"{x:.4f}")

    return positions


class Position(Resource):
    @marshal_with(pos_fields)
    @jwt_required()
    def get(self, position_id):
        uid = int(get_jwt_identity())
        user = User.get_by_id(uid)

        pos_q = PositionModel.get_user_positions(user).where(
            PositionModel.id == position_id
        )
        pos = pos_q.get_or_none()

        if pos is None:
            raise PositionDoesNotExist

        return get_positions_df(pos_q).to_dict(orient="records")[0]


class Positions(Resource):
    @marshal_with(pos_fields)
    @jwt_required()
    def get(self):
        uid = int(get_jwt_identity())
        user = User.get_by_id(uid)
        pos_q = PositionModel.get_user_positions(user)

        return get_positions_df(pos_q).to_dict("records")


class PositionsByStrategy(Resource):
    @marshal_with(by_strategy_fields)
    @jwt_required()
    def get(self, strategy_id):
        uid = int(get_jwt_identity())
        user = User.get_by_id(uid)

        try:
            strategy = Strategy.get_by_id(strategy_id)
        except Strategy.DoesNotExist:
            raise StrategyDoesNotExist

        pos_q = (
            PositionModel.get_user_positions(user)
            .where(Strategy.id == strategy_id)
            .where(PositionModel.status == "closed")
            .order_by(PositionModel.id)
        )

        pos_df = get_positions_df(pos_q)

        if pos_df.empty:
            return pos_df.to_dict("records")

        pos_df["max_equity"] = pos_df.entry_cost
        pos_df["min_equity"] = pos_df.entry_cost
        pos_df["drawdown"] = 0

        for i, row in pos_df.iterrows():
            max_drawdown = 0
            max_equity = row.max_equity
            min_equity = row.min_equity

            prices = (
                Price.get_by_strategy(strategy)
                .where(Price.time >= row.entry_time)
                .where(Price.time <= row.exit_time)
            )

            for p in prices:
                curr_equity = p.close * row.entry_amount
                curr_drawdown = (curr_equity - max_equity) / max_equity * 100

                if curr_drawdown < max_drawdown:
                    max_drawdown = curr_drawdown

                if curr_equity > max_equity:
                    max_equity = curr_equity
                    min_equity = curr_equity

                if curr_equity < min_equity:
                    min_equity = curr_equity

            quote_precision = D("10") ** (D("-1") * row.quote_precision)
            price_precision = D("10") ** (D("-1") * row.price_precision)

            max_equity = max_equity * quote_precision
            max_equity = max_equity.quantize(price_precision)

            min_equity = min_equity * quote_precision
            min_equity = min_equity.quantize(price_precision)

            pos_df.loc[i, "drawdown"] = max_drawdown
            pos_df.loc[i, "max_equity"] = max_equity
            pos_df.loc[i, "min_equity"] = min_equity

        return pos_df.to_dict("records")
